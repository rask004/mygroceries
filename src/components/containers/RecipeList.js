import React, { Component } from 'react';
import {FaTrash} from 'react-icons/fa';
import ListWithAddItemByNameForm from '../ui/ListWithAddItemByNameForm';
import PropTypes from 'prop-types';
import Button from '@material-ui/core/Button';
import ListItem from '@material-ui/core/ListItem';
import 'typeface-roboto';
import '../../css/recipemaster.css';


class RecipeList extends Component {
    handleSubmitNewRecipe = (title) => {
        const ids = this.props.recipes.map((item) => (item.id));
        const maxId = Number(Math.max(...ids));
        this.props.onAddRecipe(maxId + 1, title);
    }

    render() {
        const {
            recipes,
            currentRecipeId,
            onSelectRecipe, 
            onRemoveRecipe
        } = this.props;
        const children = recipes.map( (item) => {
            const id = Number(item.id);
            return (
                <ListItem key={id} selected={id === currentRecipeId}>
                    <Button color="default" size="medium" variant="text"
                        onClick={() => onSelectRecipe(id)}
                        href={"/recipe/" + id + "/"}
                        disableRipple={false}
                        disableFocusRipple={false}
                        className="action-select"
                    >
                        {item.title}
                    </Button>
                    <Button color="primary" size="medium" variant="contained"
                        onClick={() => onRemoveRecipe(id)}
                        disableRipple={false}
                        disableFocusRipple={false}
                        className="action-remove"
                    >
                        <FaTrash/>
                    </Button>
                </ListItem>
            );
        });
        return (
            <ListWithAddItemByNameForm orderedlist={false} placeholder="Add Recipe..."
                classFormName="recipe-add" addItem={this.handleSubmitNewRecipe}>
                {children}
            </ListWithAddItemByNameForm>
        );
    }
}

RecipeList.propTypes = {
    currentRecipeId: PropTypes.number.isRequired,
    recipes: PropTypes.array.isRequired,
    onAddRecipe: PropTypes.func.isRequired,
    onSelectRecipe: PropTypes.func.isRequired,
    onRemoveRecipe: PropTypes.func.isRequired
}


export default RecipeList;