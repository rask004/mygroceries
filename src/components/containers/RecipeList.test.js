import React from 'react';
import {create} from 'react-test-renderer';

import RecipeList from './RecipeList';

import sampleData from '../../data/initialState.json';


const recipes = sampleData.recipes;
const id = recipes[0].id;

// constant component for first testsuite.
const recipeList = create(<RecipeList
        currentRecipeId={id}
        recipes={recipes}
        onAddRecipe={jest.fn()}
        onSelectRecipe={jest.fn()}
        onRemoveRecipe={jest.fn()}
    />);


describe('InstructionsList Component render tests', () => {

    test('renders correctly', () => {
        expect(recipeList.toJSON()).toMatchSnapshot();
    });

    test('InstructionsList renders the listed instructions', () => {
        const rootInstance = recipeList.root;
        const listitems = rootInstance.findAllByType("li");
        for (let i = 0; i < listitems.length; i++) {
            const title = recipes[i].title;
            const selectButton = listitems[i].findByProps({className:"action-select"});
            expect(selectButton.props.children).toBe(title);
        }
    });
});


describe('InstructionsList Component behaviour tests', () => {

    test('add button click calls onAddRecipe', () => {
        const newRecipeTitle = "test recipe";
        const id = 4;
        const component = recipeList.root;
        const addForm = component.findByProps({className:"recipe-add"});
        const titleField = addForm.findByProps({name: "newItemName"});

        const mockChangeEvent = {
            target: {
                name:  "newItemName",
                value: newRecipeTitle
            }
        };
        titleField.props.onChange(mockChangeEvent);
        addForm.props.onSubmit({preventDefault: () => {}});

        expect(component.props.onAddRecipe).toHaveBeenCalledWith(id, newRecipeTitle);
    });

    test('title button click calls onSelectRecipe', () => {
        const selectedId = recipes[1].id;
        const component = recipeList.root;
        const listitem = component.findAllByType("li")[1];
        const selectButton = listitem.findByProps({className:"action-select"});
        selectButton.props.onClick();
        expect(component.props.onSelectRecipe).toHaveBeenCalledWith(selectedId);
    });

    test('remove button click calls onRemoveRecipe', () => {
        const component = recipeList.root;
        const listitem = component.findAllByType("li")[0];
        const id = recipes[0].id;
        const removeButton = listitem.findByProps({className:"action-remove"});
        removeButton.props.onClick();

        expect(component.props.onRemoveRecipe).toHaveBeenCalledWith(id);
    });
});
