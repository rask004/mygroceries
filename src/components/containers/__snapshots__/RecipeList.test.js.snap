// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InstructionsList Component render tests renders correctly 1`] = `
<div>
  
  <ol
    className="MuiList-root-1 MuiList-padding-2"
  >
    <li
      className="MuiListItem-root-5 MuiListItem-default-8 MuiListItem-gutters-13 MuiListItem-selected-16"
      disabled={false}
    >
      <a
        className="MuiButtonBase-root-43 MuiButton-root-17 MuiButton-text-19 MuiButton-flat-22 action-select"
        href="/recipe/1/"
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        role="button"
        tabIndex="0"
      >
        <span
          className="MuiButton-label-18"
        >
          test recipe
        </span>
        <span
          className="MuiTouchRipple-root-76"
        />
      </a>
      <button
        className="MuiButtonBase-root-43 MuiButton-root-17 MuiButton-contained-28 MuiButton-containedPrimary-29 MuiButton-raised-31 MuiButton-raisedPrimary-32 action-remove"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex="0"
        type="button"
      >
        <span
          className="MuiButton-label-18"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            strokeWidth="0"
            style={
              Object {
                "color": undefined,
              }
            }
            viewBox="0 0 448 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M432 32H312l-9.4-18.7A24 24 0 0 0 281.1 0H166.8a23.72 23.72 0 0 0-21.4 13.3L136 32H16A16 16 0 0 0 0 48v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16zM53.2 467a48 48 0 0 0 47.9 45h245.8a48 48 0 0 0 47.9-45L416 128H32z"
            />
          </svg>
        </span>
        <span
          className="MuiTouchRipple-root-76"
        />
      </button>
    </li>
    <li
      className="MuiListItem-root-5 MuiListItem-default-8 MuiListItem-gutters-13"
      disabled={false}
    >
      <a
        className="MuiButtonBase-root-43 MuiButton-root-17 MuiButton-text-19 MuiButton-flat-22 action-select"
        href="/recipe/2/"
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        role="button"
        tabIndex="0"
      >
        <span
          className="MuiButton-label-18"
        >
          another test recipe
        </span>
        <span
          className="MuiTouchRipple-root-76"
        />
      </a>
      <button
        className="MuiButtonBase-root-43 MuiButton-root-17 MuiButton-contained-28 MuiButton-containedPrimary-29 MuiButton-raised-31 MuiButton-raisedPrimary-32 action-remove"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex="0"
        type="button"
      >
        <span
          className="MuiButton-label-18"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            strokeWidth="0"
            style={
              Object {
                "color": undefined,
              }
            }
            viewBox="0 0 448 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M432 32H312l-9.4-18.7A24 24 0 0 0 281.1 0H166.8a23.72 23.72 0 0 0-21.4 13.3L136 32H16A16 16 0 0 0 0 48v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16zM53.2 467a48 48 0 0 0 47.9 45h245.8a48 48 0 0 0 47.9-45L416 128H32z"
            />
          </svg>
        </span>
        <span
          className="MuiTouchRipple-root-76"
        />
      </button>
    </li>
    <li
      className="MuiListItem-root-5 MuiListItem-default-8 MuiListItem-gutters-13"
      disabled={false}
    >
      <a
        className="MuiButtonBase-root-43 MuiButton-root-17 MuiButton-text-19 MuiButton-flat-22 action-select"
        href="/recipe/3/"
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        role="button"
        tabIndex="0"
      >
        <span
          className="MuiButton-label-18"
        >
          the third test recipe
        </span>
        <span
          className="MuiTouchRipple-root-76"
        />
      </a>
      <button
        className="MuiButtonBase-root-43 MuiButton-root-17 MuiButton-contained-28 MuiButton-containedPrimary-29 MuiButton-raised-31 MuiButton-raisedPrimary-32 action-remove"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex="0"
        type="button"
      >
        <span
          className="MuiButton-label-18"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            strokeWidth="0"
            style={
              Object {
                "color": undefined,
              }
            }
            viewBox="0 0 448 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M432 32H312l-9.4-18.7A24 24 0 0 0 281.1 0H166.8a23.72 23.72 0 0 0-21.4 13.3L136 32H16A16 16 0 0 0 0 48v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16zM53.2 467a48 48 0 0 0 47.9 45h245.8a48 48 0 0 0 47.9-45L416 128H32z"
            />
          </svg>
        </span>
        <span
          className="MuiTouchRipple-root-76"
        />
      </button>
    </li>
  </ol>
  <form
    className="recipe-add"
    onSubmit={[Function]}
  >
    <button
      className="MuiButtonBase-root-43 MuiButton-root-17 MuiButton-contained-28 MuiButton-containedPrimary-29 MuiButton-raised-31 MuiButton-raisedPrimary-32 action-add"
      disabled={false}
      onBlur={[Function]}
      onContextMenu={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex="0"
      type="submit"
    >
      <span
        className="MuiButton-label-18"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 448 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M400 32H48C21.5 32 0 53.5 0 80v352c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48V80c0-26.5-21.5-48-48-48zm-32 252c0 6.6-5.4 12-12 12h-92v92c0 6.6-5.4 12-12 12h-56c-6.6 0-12-5.4-12-12v-92H92c-6.6 0-12-5.4-12-12v-56c0-6.6 5.4-12 12-12h92v-92c0-6.6 5.4-12 12-12h56c6.6 0 12 5.4 12 12v92h92c6.6 0 12 5.4 12 12v56z"
          />
        </svg>
      </span>
      <span
        className="MuiTouchRipple-root-76"
      />
    </button>
    <div
      className="MuiInputBase-root-59 MuiInput-root-46 MuiInput-underline-50 input-new-instruction"
      onClick={[Function]}
      required={true}
    >
      <input
        className="MuiInputBase-input-69 MuiInput-input-54"
        name="newItemName"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        placeholder="Add Recipe..."
        required={true}
        type="text"
        value=""
      />
    </div>
  </form>
</div>
`;
